File: contract/runner/src/jws.ts
// The Ed25519 JWS envelope in TypeScript: contract/CONTRACT.md rules 3
// and 4, and decision 019's two shapes (no key parameter on signing, the
// key identifier inside the signed payload).
//
// Keys are built from the contract's hex encodings through JWK, because
// node:crypto has no raw Ed25519 import and JWK is the one format that
// needs no encoder of our own.

import {
  createPrivateKey,
  createPublicKey,
  sign as edSign,
  verify as edVerify,
  type KeyObject,
} from "node:crypto";
import { canonicalize } from "./jcs.ts";

export const PROTECTED_HEADER = '{"alg":"EdDSA"}';
export const KEY_ID_FIELD = "kid";

const PROTECTED_HEADER_B64 = base64url(Buffer.from(PROTECTED_HEADER, "utf8"));

export type Verdict = { valid: boolean; kid?: string; payload?: string };

// invalidVerdict is the single false verdict rule 3 fixes: exactly
// {"valid":false} and no other member. The vectors carry its serialized
// form so both languages are held to the same bytes.
export function invalidVerdict(): Verdict {
  return { valid: false };
}

// wireVerdict is the exact bytes a verdict takes on the wire: {"valid":...}
// and nothing else, which contract rule 3 requires of both verdicts
// (decision 082 fixed the true one). kid and payload stay on the Verdict for
// a caller that verified and wants them, never on the wire, mirroring the Go
// kernel's json:"-" fields. The vectors pin both serialized forms so the two
// languages are held to the same bytes.
export function wireVerdict(verdict: Verdict): string {
  return JSON.stringify({ valid: verdict.valid });
}

// Signer holds one key and reports which. Sign takes no key parameter,
// which is decision 019: the wrong key is unexpressible rather than
// rejected.
export type Signer = {
  currentKeyId: () => string;
  sign: (signingInput: Buffer) => { signature: Buffer; keyId: string };
};

export function signerFromHex(
  keyId: string,
  privateHex: string,
  publicHex: string,
): Signer {
  const key = privateKeyFromHex(privateHex, publicHex);
  return {
    currentKeyId: () => keyId,
    sign: (signingInput: Buffer) => ({
      signature: edSign(null, signingInput, key),
      keyId,
    }),
  };
}

// sign places the signer's key identifier inside the payload, canonicalizes
// the result, and returns the compact serialization.
export function sign(payload: string, signer: Signer): string {
  const keyId = signer.currentKeyId();

  // Canonicalize the RAW payload bytes first, exactly as the Go kernel's
  // withKeyID does, so this refuses every document Go refuses before it
  // inserts anything: a duplicate member name, an over-range number, an
  // unpaired surrogate. Parsing first would produce a valid signature over
  // bytes the caller never wrote, because JSON.parse silently keeps the last
  // of a duplicate key and turns an over-range number into Infinity.
  canonicalize(payload);

  const members = JSON.parse(payload) as Record<string, unknown>;
  if (
    members === null ||
    typeof members !== "object" ||
    Array.isArray(members)
  ) {
    throw new Error("payload is not a JSON object");
  }
  if (KEY_ID_FIELD in members) {
    throw new Error("payload already carries a key identifier");
  }
  members[KEY_ID_FIELD] = keyId;

  const canonical = canonicalize(JSON.stringify(members));
  const signingInput = `${PROTECTED_HEADER_B64}.${base64url(Buffer.from(canonical, "utf8"))}`;
  const { signature, keyId: usedKeyId } = signer.sign(
    Buffer.from(signingInput, "utf8"),
  );
  if (usedKeyId !== keyId) {
    throw new Error("signer used a key other than the one it reported");
  }
  return `${signingInput}.${base64url(signature)}`;
}

// verify reads the key identifier out of the signed payload, resolves it,
// and verifies. Resolving from inside the signed bytes is what makes key
// identity unforgeable.
export function verify(
  compact: string,
  resolve: (keyId: string) => KeyObject | undefined,
): Verdict {
  const parts = compact.split(".");
  if (parts.length !== 3) return invalidVerdict();

  // Rule 3: the header is compared byte for byte and never parsed.
  let header: Buffer;
  let payload: Buffer;
  let signature: Buffer;
  try {
    header = fromBase64url(parts[0]);
    payload = fromBase64url(parts[1]);
    signature = fromBase64url(parts[2]);
  } catch {
    return invalidVerdict();
  }
  if (!header.equals(Buffer.from(PROTECTED_HEADER, "utf8"))) {
    return invalidVerdict();
  }
  if (signature.length !== 64) return invalidVerdict();

  // Rule 3 again: the payload is read but never re-canonicalized, because
  // the signature covers the bytes that arrived.
  let keyId: unknown;
  try {
    keyId = (JSON.parse(payload.toString("utf8")) as Record<string, unknown>)[
      KEY_ID_FIELD
    ];
  } catch {
    return invalidVerdict();
  }
  if (typeof keyId !== "string" || keyId === "") return invalidVerdict();

  const publicKey = resolve(keyId);
  if (publicKey === undefined) return invalidVerdict();
  if (
    !edVerify(
      null,
      Buffer.from(`${parts[0]}.${parts[1]}`, "utf8"),
      publicKey,
      signature,
    )
  ) {
    return invalidVerdict();
  }
  return { valid: true, kid: keyId, payload: payload.toString("utf8") };
}

export function publicKeyFromHex(publicHex: string): KeyObject {
  return createPublicKey({
    key: {
      kty: "OKP",
      crv: "Ed25519",
      x: base64url(hexToBytes(publicHex, 32)),
    },
    format: "jwk",
  });
}

function privateKeyFromHex(privateHex: string, publicHex: string): KeyObject {
  return createPrivateKey({
    key: {
      kty: "OKP",
      crv: "Ed25519",
      d: base64url(hexToBytes(privateHex, 32)),
      x: base64url(hexToBytes(publicHex, 32)),
    },
    format: "jwk",
  });
}

// Rule 4: a private key is its 32-byte seed in hex, a public key its
// 32-byte raw value in hex.
function hexToBytes(value: string, expectedLength: number): Buffer {
  if (!/^[0-9a-f]*$/.test(value)) throw new Error("key is not lowercase hex");
  const bytes = Buffer.from(value, "hex");
  if (bytes.length !== expectedLength) {
    throw new Error(`key is ${bytes.length} bytes, want ${expectedLength}`);
  }
  return bytes;
}

function base64url(bytes: Buffer): string {
  return bytes.toString("base64url");
}

function fromBase64url(segment: string): Buffer {
  const bytes = Buffer.from(segment, "base64url");
  // Buffer.from is lenient, so a segment that does not round-trip is
  // rejected here rather than silently accepted as a shorter value.
  if (base64url(bytes) !== segment) throw new Error("segment is not base64url");
  return bytes;
}

File: core/kernel/jws.go
package kernel

import (
	"bytes"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

// ProtectedHeader is the whole protected header, frozen by contract rule 3.
// Verification compares these bytes and does not parse them, which is why
// there is nowhere in the header for a key identifier (decision 019).
const ProtectedHeader = `{"alg":"EdDSA"}`

// KeyIDField names the key identifier inside the signed payload. Decision
// 019 put it there; the field name follows the repo's existing term for it
// (model/attestation-interface.md and the design tree call an attestation
// kid-bound) and is a single constant so the pending interface amendment
// decision 019 flagged is a one-line change here.
const KeyIDField = "kid"

// protectedHeaderB64 is the first segment of every compact serialization
// this kernel produces or accepts.
var protectedHeaderB64 = base64.RawURLEncoding.EncodeToString([]byte(ProtectedHeader))

// Errors the signing path reports.
var (
	ErrPayloadNotObject  = errors.New("kernel: payload is not a JSON object")
	ErrKeyIDPresent      = errors.New("kernel: payload already carries a key identifier")
	ErrKeyRotatedMidSign = errors.New("kernel: signer used a key other than the one it reported")
)

// Signer is the operator key provider's signing seam. trust-kernel/02
// supplies the real implementation (software locally and in CI, KMS in
// production); core/kernel/fixture supplies the deterministic one the
// golden vectors are generated with.
//
// Sign takes no key parameter, which is decision 019: key selection and
// rotation live inside the provider, so signing with another party's key
// is unexpressible rather than rejected by a runtime check somebody can
// later regress.
type Signer interface {
	// CurrentKeyID reports the key Sign would use now. Sign cannot report
	// it first, because the identifier has to be inside the payload before
	// the payload is signed.
	CurrentKeyID() (string, error)

	// Sign signs signingInput with the provider's current active key and
	// reports which key that was.
	Sign(signingInput []byte) (signature []byte, keyID string, err error)
}

// Sign places the signer's current key identifier inside the payload,
// canonicalizes the result, and returns the JWS compact serialization.
//
// The identifier goes in before the signature is computed, so altering it
// afterwards changes the signed bytes and the signature no longer verifies.
// That is the whole point of decision 019, and contract/vectors' tampered
// cases are what hold it.
func Sign(payload []byte, signer Signer) (string, error) {
	keyID, err := signer.CurrentKeyID()
	if err != nil {
		return "", err
	}

	bound, err := withKeyID(payload, keyID)
	if err != nil {
		return "", err
	}
	canonical, err := Canonicalize(bound)
	if err != nil {
		return "", err
	}

	signingInput := protectedHeaderB64 + "." + base64.RawURLEncoding.EncodeToString(canonical)
	signature, usedKeyID, err := signer.Sign([]byte(signingInput))
	if err != nil {
		return "", err
	}
	// A rotation between the two provider calls would sign with a key the
	// payload does not name, producing a record that can never verify.
	// Enforced here, and by the fixture signer's rotation test.
	if usedKeyID != keyID {
		return "", fmt.Errorf("%w: reported %q, embedded %q", ErrKeyRotatedMidSign, usedKeyID, keyID)
	}

	return signingInput + "." + base64.RawURLEncoding.EncodeToString(signature), nil
}

// withKeyID returns the payload with the key identifier added as a
// top-level member. A payload that already carries one is refused rather
// than overwritten: it is either a re-signing attempt or a caller trying to
// choose a key through the payload, and neither is a thing this kernel does.
func withKeyID(payload []byte, keyID string) ([]byte, error) {
	// Canonicalize first, so a duplicate member or a malformed document is
	// refused here rather than silently collapsed by the map below.
	if _, err := Canonicalize(payload); err != nil {
		return nil, err
	}
	var members map[string]json.RawMessage
	if err := json.Unmarshal(payload, &members); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrPayloadNotObject, err)
	}
	if _, present := members[KeyIDField]; present {
		return nil, ErrKeyIDPresent
	}
	encoded, err := json.Marshal(keyID)
	if err != nil {
		return nil, err
	}
	members[KeyIDField] = encoded
	// Marshalling a map reorders members; Canonicalize is what fixes order,
	// so this intermediate form only has to be valid JSON.
	return json.Marshal(members)
}

// Verdict is the outcome of verification. A false verdict serializes to
// exactly {"valid":false} with no other member, which is contract rule 3;
// the omitempty tags are what make that true, and the vector runners
// compare the serialized form rather than the struct.
type Verdict struct {
	Valid   bool            `json:"valid"`
	KeyID   string          `json:"kid,omitempty"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

// Invalid is the single false verdict.
func Invalid() Verdict { return Verdict{Valid: false} }

// Resolver maps a key identifier to the public key it names. party-registry
// supplies the real one, reading the registry; core/kernel/fixture supplies
// the one the vectors use.
type Resolver interface {
	PublicKey(keyID string) (ed25519.PublicKey, bool)
}

// Verify reads the key identifier out of the signed payload, resolves it,
// and verifies the signature. Reading the identifier from inside the signed
// bytes is what makes it unforgeable: a record cannot claim a key it was
// not signed with.
func Verify(compact string, resolver Resolver) Verdict {
	payload, keyID, signingInput, signature, ok := split(compact)
	if !ok {
		return Invalid()
	}
	publicKey, found := resolver.PublicKey(keyID)
	if !found {
		return Invalid()
	}
	if !ed25519.Verify(publicKey, signingInput, signature) {
		return Invalid()
	}
	return Verdict{Valid: true, KeyID: keyID, Payload: payload}
}

// VerifyWithKey verifies against one public key given directly, the shape
// decision 019 keeps separate from signing: verification names its key,
// signing never does. The payload's key identifier still has to match the
// key offered, or a record signed by one party could be presented as
// another's.
func VerifyWithKey(compact string, publicKey ed25519.PublicKey, keyID string) Verdict {
	payload, embeddedKeyID, signingInput, signature, ok := split(compact)
	if !ok || embeddedKeyID != keyID {
		return Invalid()
	}
	if !ed25519.Verify(publicKey, signingInput, signature) {
		return Invalid()
	}
	return Verdict{Valid: true, KeyID: keyID, Payload: payload}
}

// split takes a compact serialization apart and enforces contract rule 3's
// header check: the protected header is compared byte-for-byte against the
// frozen bytes and never parsed, so a header carrying an extra member is
// rejected rather than tolerated. The payload is decoded and read but never
// re-canonicalized, also rule 3: the signature covers the bytes that were
// transmitted, so re-deriving them would verify a different document.
func split(compact string) (payload json.RawMessage, keyID string, signingInput, signature []byte, ok bool) {
	parts := strings.Split(compact, ".")
	if len(parts) != 3 {
		return nil, "", nil, nil, false
	}
	header, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil || !bytes.Equal(header, []byte(ProtectedHeader)) {
		return nil, "", nil, nil, false
	}
	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, "", nil, nil, false
	}
	signature, err = base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil || len(signature) != ed25519.SignatureSize {
		return nil, "", nil, nil, false
	}

	var members map[string]json.RawMessage
	if err := json.Unmarshal(payloadBytes, &members); err != nil {
		return nil, "", nil, nil, false
	}
	if err := json.Unmarshal(members[KeyIDField], &keyID); err != nil || keyID == "" {
		return nil, "", nil, nil, false
	}
	return payloadBytes, keyID, []byte(parts[0] + "." + parts[1]), signature, true
}
